### PR TITLE
Move scalar file configs to YAMLs and use a recon scalar workflow

### PR DESCRIPTION
**Pull request by tsalo**
Closes none.

## Changes proposed in this pull request

- Move scalar file config dictionaries into YAML files (`data/scalars/*.yaml`).
- Replace `ReconScalarsDataSink` with a workflow that uses `DerivativesDataSink`.
- Pass along metadata from scalar file config dictionaries to recon map sidecars.

### CODE REVIEW

**Comment by tsalo** [comment]
@mattcieslak I've only applied the new workflow to AMICO so far, but I can update the other recon pipelines as well as along as you like the changes.

File: qsirecon/workflows/recon/utils.py [hunk 1]
```diff
@@ -82,3 +84,54 @@ def init_discard_repeated_samples_wf(
...
+        name="inputnode",
+    )
+    outputnode = pe.Node(
+        niu.IdentityInterface(fields=["scalar_files"]),
+        name="outputnode",
+    )
+    workflow = pe.Workflow(name=name)
+
+    organize_scalar_data = pe.MapNode(
+        OrganizeScalarData(),
```
Comment (mattcieslak) [inline qsirecon/workflows/recon/utils.py:109]: so this is how you get the data added to sidecars?

**Review by mattcieslak — APPROVED** [review]
I know there are still a few workflows to do, but this looks awesome. Adding in those sidecars is a huge deal

File: qsirecon/interfaces/recon_scalars.py [hunk 2]
```diff
@@ -189,45 +199,7 @@ def _run_interface(self, runtime):
...
-    "ngpar_file": {
-        "desc": "Non-Gaussianity parallel from MAPMRI",
-        "bids": {"param": "NGpar", "model": "mapmri"},
-    },
-    "ngperp_file": {
-        "desc": "Non-Gaussianity perpendicular from MAPMRI",
-        "bids": {"param": "NGperp", "model": "mapmri"},
-    },
-}
+tortoise_scalars = load_yaml(load_data("scalars/tortoise.yaml"))
```
[the remark this answers is not in the document]
Comment (tsalo) [reply qsirecon/interfaces/recon_scalars.py:202]: Thanks!

File: qsirecon/interfaces/recon_scalars.py [hunk 3]
```diff
@@ -244,22 +216,7 @@ class TORTOISEReconScalars(ReconScalars):
...
-        "bids": {"param": "isovf", "model": "noddi"},
-    },
-    "od_image": {"desc": "OD from NODDI", "bids": {"param": "od", "model": "noddi"}},
-    "directions_image": {
-        "desc": "Peak directions from NODDI",
-        "reorient_on_resample": True,
-        "bids": {"param": "direction", "model": "noddi"},
-    },
-}
+amico_scalars = load_yaml(load_data("scalars/amico.yaml"))
```
[the remark this answers is not in the document]
Comment (tsalo) [reply qsirecon/interfaces/recon_scalars.py:219]: Can do

File: qsirecon/interfaces/recon_scalars.py [hunk 4]
```diff
@@ -276,33 +233,7 @@ class AMICOReconScalars(ReconScalars):
...
-    "md_file": {"desc": "Mean Diffusivity", "bids": {"param": "md", "model": "tensor"}},
-    "ad_file": {"desc": "AD", "bids": {"param": "ad", "model": "tensor"}},
-    "rd_file": {"desc": "Radial Diffusivity", "bids": {"param": "rd", "model": "tensor"}},
-    "gfa_file": {
-        "desc": "Generalized Fractional Anisotropy",
-        "bids": {"model": "GQI", "param": "gfa"},
-    },
-    "iso_file": {"desc": "Isotropic Diffusion", "bids": {"model": "GQI", "param": "iso"}},
-}
+dsistudio_scalars = load_yaml(load_data("scalars/dsistudio.yaml"))
```
[the remark this answers is not in the document]
Comment (tsalo) [reply qsirecon/interfaces/recon_scalars.py:236]: Can do

File: qsirecon/workflows/recon/utils.py (hunk 1 above)
Comment (tsalo) [reply qsirecon/workflows/recon/utils.py:109]: It splits up the scalar dictionary object into individual fields that the DerivativesDataSink can work with, including individual entities that go into the filename and the metadata that goes into the sidecar.

If we want to patch in other metadata (e.g., `Sources` per #69) then we'll need to merge the metadata dictionary from `OrganizeScalarData` with whatever those metadata are before patching them into the DerivativesDataSink.

**Comment by tsalo** [comment]
I'll start changing over the other pipelines then. Thanks!

File: qsirecon/workflows/base.py [hunk 5]
```diff
@@ -355,10 +355,31 @@ def _load_recon_spec(spec_name):
...
+                if "qsirecon_suffix" not in new_node.keys():
+                    # Infer the suffix from the source node
+                    for temp_node in spec["nodes"]:
+                        if temp_node["name"] == scalar_source:
+                            new_node["qsirecon_suffix"] = temp_node["qsirecon_suffix"]
+                            continue
+
+                spec["nodes"].append(new_node)
+        else:
+            spec["nodes"].append(node)
```
Comment (tsalo) [inline qsirecon/workflows/base.py:381]: @mattcieslak check this out. WDYT?

Comment (mattcieslak) [reply qsirecon/workflows/base.py:381]: wow this is way prettier than I thought it would have to be

Comment (tsalo) [reply qsirecon/workflows/base.py:381]: As long as nodes with `scalars_from` are endpoints in the workflow (i.e., don't feed into other nodes), I think it should work. If they do feed into others then I might need to make it a bit more complicated.

Comment (mattcieslak) [reply qsirecon/workflows/base.py:381]: does .copy() work here? I remember needing to deepcopy nipype objects back in the day

Comment (tsalo) [reply qsirecon/workflows/base.py:381]: It should since this is just a dictionary at the moment.

Comment (tsalo) [reply qsirecon/workflows/base.py]: Switched to deepcopy to be safe given our convo in scrum.

File: qsirecon/tests/data/scalar_mapper_outputs.txt
[the remark this answers is not in the document]
Comment (mattcieslak) [reply qsirecon/tests/data/scalar_mapper_outputs.txt:1]: it looks like it's writing out all the model outputs to all the qsirecon suffix directories.

File: qsirecon/tests/data/scalar_mapper_outputs.txt [hunk 6]
```diff
@@ -10,24 +10,74 @@ derivatives/qsirecon-DIPYDKI/logs/CITATION.tex
 derivatives/qsirecon-DIPYDKI/sub-ABCD
 derivatives/qsirecon-DIPYDKI/sub-ABCD.html
 derivatives/qsirecon-DIPYDKI/sub-ABCD/dwi
+derivatives/qsirecon-DIPYDKI/sub-ABCD/dwi/sub-ABCD_acq-10per000_space-MNI152NLin2009cAsym_model-GQI_param-gfa_dwimap.json
+derivatives/qsirecon-DIPYDKI/sub-ABCD/dwi/sub-ABCD_acq-10per000_space-MNI152NLin2009cAsym_model-GQI_param-gfa_dwimap.nii.gz
+derivatives/qsirecon-DIPYDKI/sub-ABCD/dwi/sub-ABCD_acq-10per000_space-MNI152NLin2009cAsym_model-GQI_param-iso_dwimap.json
+derivatives/qsirecon-DIPYDKI/sub-ABCD/dwi/sub-ABCD_acq-10per000_space-MNI152NLin2009cAsym_model-GQI_param-iso_dwimap.nii.gz
+derivatives/qsirecon-DIPYDKI/sub-ABCD/dwi/sub-ABCD_acq-10per000_space-MNI152NLin2009cAsym_model-GQI_param-qa_dwimap.json
```
Comment (mattcieslak) [inline qsirecon/tests/data/scalar_mapper_outputs.txt:17]: these should be in the dsistudio qsirecon suffix

Comment (tsalo) [reply qsirecon/tests/data/scalar_mapper_outputs.txt:17]: Thanks for catching that. I'll look into it.

Comment (tsalo) [reply qsirecon/tests/data/scalar_mapper_outputs.txt:17]: I think `scalar_gatherer` is the culprit

Comment (mattcieslak) [reply qsirecon/tests/data/scalar_mapper_outputs.txt:17]: ah! we might not need scalar_gatherer at all now that the scalar mappers are being duplicated for each mapping

Comment (tsalo) [reply qsirecon/tests/data/scalar_mapper_outputs.txt]: Fixed!

**Pull request merged**